### PR TITLE
adapter: Remove unwrap()

### LINKED ISSUE #13306: Replace `unwrap()` with `expect()` in the Coordinator
In `coord.rs` there are 79 calls to `unwrap()`. I think we should prefer `expect()` with a clear explanation of why it will never fail or why failure should cause a panic. The vast majority of these are calls to STORAGE/COMPUTE. Refactoring the `unwrap`s/`expect`s to helper methods would help reduce the total number and de-duplicate explanation strings in `expect`.

### PR DESCRIPTION
**Pull request by jkosh44**
This commit removes all uses of `unwrap()` from the mz-adapter crate, and replaces it with `expect()` or `unwrap_or_terminate()`.

Fixes #13306

### Motivation
This PR adds a known-desirable feature.

### Checklist

- [X] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/adapter/src/util.rs [hunk 1]
```diff
@@ -285,3 +297,77 @@ impl ShouldHalt for StashError {
...
+        }
+    }
+}
+
+// Needed for some functions that return an () error. We have no info, so just panic.
+impl ShouldHalt for () {
+    fn should_halt(&self) -> bool {
+        false
+    }
+}
```
Comment (jkosh44) [inline src/adapter/src/util.rs:373]: @benesch No need to review this entire PR, but I was hoping that you could spot check these. As a rule, I called `unwrap_or_terminate` for all controller methods. The logic being, that even if some of them return no halt-able errors now, they may return halt-able errors in the future. That's why we have some implementations always returning `false`.

File: src/adapter/src/coord.rs [hunk 2]
```diff
@@ -1210,15 +1211,19 @@ pub async fn serve(
...
-                aws_external_id_prefix: connection_context.aws_external_id_prefix.clone().unwrap(),
-            })
-        } else {
-            None
-        };
+    let aws_principal_context = if aws_account_id.is_some()
+        && connection_context.aws_external_id_prefix.is_some()
+    {
+        Some(AwsPrincipalContext {
+            aws_account_id: aws_account_id.expect("known to be `Some` from`is_some()` call above"),
```
Comment (mjibson) [inline src/adapter/src/coord.rs:1218]:
```suggestion
            aws_account_id: aws_account_id.expect("known to be `Some` from `is_some()` call above"),
```

File: src/adapter/src/coord.rs [hunk 3]
```diff
@@ -1210,15 +1211,19 @@ pub async fn serve(
...
+            aws_external_id_prefix: connection_context
+                .aws_external_id_prefix
+                .clone()
+                .expect("known to be `Some` from`is_some()` call above"),
```
Comment (mjibson) [inline src/adapter/src/coord.rs:1222]:
```suggestion
                .expect("known to be `Some` from `is_some()` call above"),
```

**Review by mjibson — APPROVED** [review]

File: src/adapter/src/util.rs [hunk 4]
```diff
@@ -285,3 +297,70 @@ impl ShouldHalt for StashError {
...
+
+impl ShouldHalt for TransformError {
+    fn should_halt(&self) -> bool {
+        match self {
+            TransformError::Internal(_)
+            | TransformError::LetRecUnsupported
+            | TransformError::IdentifierMissing(_) => false,
+        }
+    }
+}
```
Comment (jkosh44) [inline src/adapter/src/util.rs:366]: Just re-posting my previous comment on non-outdated code:
> @benesch No need to review this entire PR, but I was hoping that you could spot check these. As a rule, I called `unwrap_or_terminate` for all controller methods. The logic being, that even if some of them return no halt-able errors now, they may return halt-able errors in the future. That's why we have some implementations always returning `false`.

Comment (benesch) [reply src/adapter/src/util.rs:366]: Makes sense to me!

**Pull request merged**